domain/port/state/state.go

```diff
@@ -0,0 +1,269 @@
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing get opened ports statement")
+	}
+
+	results := []endpointPortRange{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, query, unitUUID).GetAll(&results)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting opened ports for unit %q", unit)
+	}
+
+	groupedPortRanges := network.GroupedPortRanges{}
+	for _, endpointPortRange := range results {
+		if _, ok := groupedPortRanges[endpointPortRange.Endpoint]; !ok {
+			groupedPortRanges[endpointPortRange.Endpoint] = []network.PortRange{}
+		}
+		groupedPortRanges[endpointPortRange.Endpoint] = append(groupedPortRanges[endpointPortRange.Endpoint], endpointPortRange.decode())
+	}
+	return groupedPortRanges, nil
+}
+
+// UpdateUnitPorts opens and closes ports for the endpoints of a given unit.
+// The service layer must ensure that opened and closed ports for the same
+// endpoints must not conflict.
+func (st *State) UpdateUnitPorts(ctx context.Context, unit string, openPorts, closePorts network.GroupedPortRanges) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	endpointsUnderActionSet := set.NewStrings()
+	for endpoint := range openPorts {
+		endpointsUnderActionSet.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		endpointsUnderActionSet.Add(endpoint)
+	}
+	endpointsUnderAction := endpoints(endpointsUnderActionSet.Values())
+
+	unitUUID := unitUUID{UUID: unit}
+
+	getOpenedPortsOnEndpoints, err := st.Prepare(`
+SELECT &endpointPortRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
+WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
+AND unit_endpoint.endpoint IN ($endpoints[:])
+`, endpointPortRange{}, unitUUID, endpointsUnderAction)
+	if err != nil {
+		return errors.Annotate(err, "preparing get opened ports on endpoints statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare("DELETE FROM port_range WHERE unit_endpoint_uuid IN ($endpointUUIDs[:]) ", endpointUUIDs{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		endpoints, err := st.ensureEndpoints(ctx, tx, unitUUID, endpointsUnderAction)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		protocolMap, err := st.getProtocolMap(ctx, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Get the current opened ports for the unit's endpoints.
+		// This will be used to calculate the changes required.
+		var currentOpenedEndpointPorts []endpointPortRange
```

Comment:
I know you've already done a load of splitting up of this function, but it might be quite nice to split it up even more, to the point where you can easily read a sequence of calls+error checks and this master function doesn't actually contain any of the nitty gritty struct transformation stuff.

Could be a bit like simon did [here](https://github.com/juju/juju/pull/17865/files#diff-7961d740899680f58e70493a1801eb5c86cf80e665b8615b876d22e0dd14b7faR262). It makes it super readable and grokkable. Of course that link is a bit simpler than this case, but it would be good if it was closer to that.